src/layers/SoftMax.jl

```diff
@@ -1,41 +1,66 @@
 type SoftMax <: Nonlinearity
-    last_input  :: Array{Float64}
-    last_output :: Array{Float64}
-    last_loss   :: Array{Float64}
+    x  :: Array{Float64}
+    y :: Array{Float64}
+    has_init :: Bool
+    jacobian :: Array{Float64}
+    dldx :: Array{Float64}
+    # last_loss   :: Array{Float64}
 
     function SoftMax()
-        return new(Float64[], Float64[], Float64[])
+        return new(Float64[], Float64[])
     end
 end
 
-function forward(l::SoftMax, X::Array{Float64}; kwargs...)
-	X = broadcast(+, X, -maximum(X))
-    l.last_input = X
-    sumX = sum(exp(X))
-    Y = (x ./ sumX)
-    l.last_output = Y
-    return Y
+function init(l::SoftMax, p::Union{Layer,Void}, config::Dict{String,Any}; kwargs...)
+  if p == nothing
+      @assert ndims(config["input_size"]) == 1 # TODO: maybe an error message?
+      out_size = (config["batch_size"], config["input_size"][1])
+  else
+      out_size = getOutputSize(p)
+  end
+  l.x = Array{Float64}(out_size)
+  l.y = Array{Float64}(out_size)
+  l.has_init = true;
+  m,n = out_size
+  l.jacobian = Array{Float64}(n,n)
+  l.dldx = Array{Float64}(out_size)
+end
+
+function forward(l::SoftMax, X::Array{Float64,2}; kwargs...)
 
+    l.x = X
+    # iterating each row/picture
+    lxexp = exp(l.x)
```

Comment:
Reimplemented `SoftMax` forward as mention in other replies. Not sure if they are `in-place` as you expect.